tutorials/Tutorial17_GPL.py

```diff
@@ -0,0 +1,249 @@
+    document_store.write_documents([{"content": t} for t in corpus])
+
+    retriever = EmbeddingRetriever(
+        document_store=document_store,
+        embedding_model="sentence-transformers/msmarco-distilbert-base-tas-b",
+        model_format="sentence_transformers",
+        max_seq_len=max_seq_length,
+        progress_bar=False,
+    )
+    document_store.update_embeddings(retriever)
+
+    """## Optionally download pre-generated questions or even generate them outside of Haystack
+    
+    The first step of the GPL algorithm requires us to generate questions for a given text passage. Even though our pre-COVID trained model hasn't seen any COVID-related content, it can still produce sensible queries by copying words from the input text. As generating questions from 10k documents is a bit slow (depending on GPU used), we'll download question/document pairs directly from the HuggingFace hub.
+    
+    """
+
+    from tqdm.auto import tqdm
+
+    query_doc_pairs = []
+
+    load_queries_from_hub = True
+
+    # Generation of the queries is quite slow in Colab due to the old GPU and the limited CPU
+    # I pre-computed the queries and uploaded these to the HF dataset hub. Here we just download them
+    if load_queries_from_hub:
+        generated_queries = load_dataset("nreimers/trec-covid-generated-queries", split="train")
+        for row in generated_queries:
+            query_doc_pairs.append({"question": row["query"], "document": row["doc"]})
+    else:
+        # Load doc2query model
+        t5_name = "doc2query/msmarco-t5-base-v1"
+        t5_tokenizer = AutoTokenizer.from_pretrained(t5_name)
+        t5_model = AutoModelForSeq2SeqLM.from_pretrained(t5_name).cuda()
+
+        batch_size = 32
+        queries_per_doc = 3
+
+        for start_idx in tqdm(range(0, len(corpus), batch_size)):
+            corpus_batch = corpus[start_idx : start_idx + batch_size]
+            enc_inp = t5_tokenizer(
+                corpus_batch, max_length=max_seq_length, truncation=True, padding=True, return_tensors="pt"
+            )
+
+            outputs = t5_model.generate(
+                input_ids=enc_inp["input_ids"].cuda(),
+                attention_mask=enc_inp["attention_mask"].cuda(),
+                max_length=64,
+                do_sample=True,
+                top_p=0.95,
+                num_return_sequences=queries_per_doc,
+            )
+
+            decoded_output = t5_tokenizer.batch_decode(outputs, skip_special_tokens=True)
+
+            for idx, query in enumerate(decoded_output):
+                corpus_id = int(idx / queries_per_doc)
+                query_doc_pairs.append({"question": query, "document": corpus_batch[corpus_id]})
+
+    print("Generated queries:", len(query_doc_pairs))
+
+    """# Use PseudoLabelGenerator to genenerate Retriever adaptation training data
+    
+    PseudoLabelGenerator#run will execute all three steps of the GPL [algorithm](https://github.com/UKPLab/gpl#how-does-gpl-work):
```

Comment:
I think # before "run" is not needed here.

Reply:
Ok, sure.